File: shiden/scripts/export-from-neo4j.js
#!/usr/bin/env node
/**
 * Neo4jから教育理論データをエクスポートするスクリプト
 * TSK-001: Neo4jからデータエクスポート
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const OUTPUT_DIR = path.join(__dirname, '..', 'src', 'data');

// 出力ディレクトリ作成
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

console.log('📥 Neo4jから教育理論データをエクスポート中...\n');

// 理論データをエクスポート
console.log('1. 理論データを取得中...');
const theoriesQuery = `
MATCH (t:Theory) 
RETURN {
  id: t.id,
  name: t.name,
  name_ja: t.name_ja,
  category: t.category,
  description: t.description,
  description_ja: t.description_ja,
  theorists: COALESCE(t.theorists, []),
  priority: COALESCE(t.priority, 5),
  key_principles: COALESCE(t.key_principles, []),
  applications: COALESCE(t.applications, []),
  strengths: COALESCE(t.strengths, []),
  limitations: COALESCE(t.limitations, [])
} as theory
`.replace(/\n/g, ' ').trim();

try {
  const theoriesRaw = execSync(
    `docker exec tenjin-neo4j cypher-shell -u neo4j -p password "${theoriesQuery}" --format plain 2>/dev/null`,
    { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 }
  );

  // パース処理
  const lines = theoriesRaw.split('\n').filter(line => line.trim() && line.trim() !== 'theory');
  const theories = [];

  for (const line of lines) {
    try {
      // Neo4jの出力形式を修正
      let jsonStr = line.trim()
        .replace(/NULL/g, 'null')
        .replace(/(\w+):/g, '"$1":')  // キーをダブルクォート
        .replace(/'/g, '"');           // シングルクォートをダブルクォートに
      
      const theory = JSON.parse(jsonStr);
      theories.push(theory);
    } catch (e) {
      // 単純なJSONパースが失敗した場合、手動パース
      console.log(`  ⚠️ パースエラー (スキップ): ${line.substring(0, 50)}...`);
    }
  }

  console.log(`  ✅ ${theories.length}件の理論を取得\n`);

  // 関係データをエクスポート
  console.log('2. 関係データを取得中...');
  const relationsQuery = `
MATCH (t1:Theory)-[r]->(t2:Theory) 
RETURN {
  source_id: t1.id,
  target_id: t2.id,
  relation_type: type(r),
  weight: COALESCE(r.weight, 1.0)
} as relation
`.replace(/\n/g, ' ').trim();

  const relationsRaw = execSync(
    `docker exec tenjin-neo4j cypher-shell -u neo4j -p password "${relationsQuery}" --format plain 2>/dev/null`,
    { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 }
  );

  const relLines = relationsRaw.split('\n').filter(line => line.trim() && line.trim() !== 'relation');
  const relations = [];

  for (const line of relLines) {
    try {
      let jsonStr = line.trim()
        .replace(/NULL/g, 'null')
        .replace(/(\w+):/g, '"$1":')
        .replace(/'/g, '"');
      
      const relation = JSON.parse(jsonStr);
      relations.push(relation);
    } catch (e) {
      // スキップ
    }
  }

  console.log(`  ✅ ${relations.length}件の関係を取得\n`);

  // ファイル出力
  const theoriesPath = path.join(OUTPUT_DIR, 'theories.json');
  const relationsPath = path.join(OUTPUT_DIR, 'relations.json');

  fs.writeFileSync(theoriesPath, JSON.stringify(theories, null, 2));
  fs.writeFileSync(relationsPath, JSON.stringify(relations, null, 2));

  console.log('3. 出力完了:');
  console.log(`  📄 ${theoriesPath}`);
  console.log(`  📄 ${relationsPath}`);
  console.log('\n✅ エクスポート完了!');

} catch (error) {
  console.error('❌ エラー:', error.message);
  process.exit(1);
}

File: shiden/scripts/export-from-neo4j.py
#!/usr/bin/env python3
"""
Neo4jから教育理論データをエクスポートするスクリプト
TSK-001: Neo4jからデータエクスポート
"""

import json
import os
from neo4j import GraphDatabase

# Neo4j接続設定
URI = "bolt://localhost:7687"
USER = "neo4j"
PASSWORD = "password"

OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "..", "src", "data")

def export_theories(driver):
    """理論データをエクスポート"""
    theories = []
    
    with driver.session() as session:
        result = session.run("""
            MATCH (t:Theory)
            RETURN t.id as id,
                   t.name as name,
                   t.name_ja as name_ja,
                   t.category as category,
                   t.description as description,
                   t.description_ja as description_ja,
                   t.theorists as theorists,
                   t.priority as priority,
                   t.key_principles as key_principles,
                   t.applications as applications,
                   t.strengths as strengths,
                   t.limitations as limitations
            ORDER BY t.id
        """)
        
        for record in result:
            theory = {
                "id": record["id"],
                "name": record["name"],
                "name_ja": record["name_ja"],
                "category": record["category"],
                "description": record["description"],
                "description_ja": record["description_ja"],
                "theorists": record["theorists"] or [],
                "priority": record["priority"] or 5,
                "key_principles": record["key_principles"] or [],
                "applications": record["applications"] or [],
                "strengths": record["strengths"] or [],
                "limitations": record["limitations"] or []
            }
            theories.append(theory)
    
    return theories

def export_relations(driver):
    """関係データをエクスポート"""
    relations = []
    
    with driver.session() as session:
        result = session.run("""
            MATCH (t1:Theory)-[r]->(t2:Theory)
            RETURN t1.id as source_id,
                   t2.id as target_id,
                   type(r) as relation_type,
                   COALESCE(r.weight, 1.0) as weight
            ORDER BY t1.id, t2.id
        """)
        
        for record in result:
            relation = {
                "source_id": record["source_id"],
                "target_id": record["target_id"],
                "relation_type": record["relation_type"].lower(),
                "weight": record["weight"]
            }
            relations.append(relation)
    
    return relations

def main():
    print("📥 Neo4jから教育理論データをエクスポート中...\n")
    
    # 出力ディレクトリ作成
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    # Neo4j接続
    driver = GraphDatabase.driver(URI, auth=(USER, PASSWORD))
    
    try:
        # 理論データ
        print("1. 理論データを取得中...")
        theories = export_theories(driver)
        print(f"  ✅ {len(theories)}件の理論を取得\n")
        
        # 関係データ
        print("2. 関係データを取得中...")
        relations = export_relations(driver)
        print(f"  ✅ {len(relations)}件の関係を取得\n")
        
        # ファイル出力
        theories_path = os.path.join(OUTPUT_DIR, "theories.json")
        relations_path = os.path.join(OUTPUT_DIR, "relations.json")
        
        with open(theories_path, "w", encoding="utf-8") as f:
            json.dump(theories, f, ensure_ascii=False, indent=2)
        
        with open(relations_path, "w", encoding="utf-8") as f:
            json.dump(relations, f, ensure_ascii=False, indent=2)
        
        print("3. 出力完了:")
        print(f"  📄 {theories_path}")
        print(f"  📄 {relations_path}")
        print("\n✅ エクスポート完了!")
        
    finally:
        driver.close()

if __name__ == "__main__":
    main()
